Tools/cases_generator/parser.py

```diff
@@ -0,0 +1,222 @@
+                            if self.expect(lx.RBRACKET):
+                                return f"{tkn.text}[{arg.text}*{num.text}]"
+                raise self.make_syntax_error("Expected argument in brackets", tkn)
+
+            return tkn.text
+        if self.expect(lx.CONDOP):
+            while self.expect(lx.CONDOP):
+                pass
+            return "??"
+        return None
+
+    def outputs(self):
+        # output (, output)*
+        here = self.getpos()
+        if outp := self.output():
+            near = self.getpos()
+            if self.expect(lx.COMMA):
+                if rest := self.outputs():
+                    return [outp] + rest
+            self.setpos(near)
+            return [outp]
+        self.setpos(here)
+        return None
+
+    def output(self):
+        return self.input()  # TODO: They're not quite the same.
+
+    @contextual
+    def family_def(self) -> Family | None:
+        here = self.getpos()
+        if (tkn := self.expect(lx.IDENTIFIER)) and tkn.text == "family":
+            if self.expect(lx.LPAREN):
+                if (tkn := self.expect(lx.IDENTIFIER)):
+                    name = tkn.text
+                    if self.expect(lx.RPAREN):
+                        if self.expect(lx.EQUALS):
+                            if members := self.members():
+                                if self.expect(lx.SEMI):
+                                    return Family(name, members)
+        return None
+
+    def members(self):
+        here = self.getpos()
+        if tkn := self.expect(lx.IDENTIFIER):
+            near = self.getpos()
+            if self.expect(lx.COMMA):
+                if rest := self.members():
+                    return [tkn.text] + rest
+            self.setpos(near)
+            return [tkn.text]
+        self.setpos(here)
+        return None
+
+    @contextual
+    def block(self) -> Block:
+        tokens = self.c_blob()
+        return Block(tokens)
+
+    def c_blob(self):
+        tokens = []
+        level = 0
+        while tkn := self.next(raw=True):
+            if tkn.kind in (lx.LBRACE, lx.LPAREN, lx.LBRACKET):
+                level += 1
+            elif tkn.kind in (lx.RBRACE, lx.RPAREN, lx.RBRACKET):
+                level -= 1
+                if level <= 0:
+                    break
+            tokens.append(tkn)
+        return tokens
+
+
+if __name__ == "__main__":
+    import sys
+    if sys.argv[1:]:
+        filename = sys.argv[1]
+        if filename == "-c" and sys.argv[2:]:
+            src = sys.argv[2]
+            filename = None
+        else:
+            with open(filename) as f:
+                src = f.read()
+            srclines = src.splitlines()
+            begin = srclines.index("// BEGIN BYTECODES //")
+            end = srclines.index("// END BYTECODES //")
+            src = "\n".join(srclines[begin+1 : end])
+    else:
+        filename = None
+        src = "if (x) { x.foo; // comment\n}"
```

Comment:
Not sure why this is here... maybe leftover debugging?

We'd probably want to just raise, right?

Reply:
This entire `__main__` block is for debugging the parser, and this is some sample test input. :-)